src/include/ipc/topology.h

```diff
@@ -48,6 +48,7 @@ enum sof_comp_type {
 	SOF_COMP_DCBLOCK,
 	SOF_COMP_SMART_AMP,		/**< smart amplifier component */
 	SOF_COMP_MODULE_ADAPTER,		/**< module adapter */
+	SOF_COMP_DYNAMIC,		/**< dynamic lodable library */
```

Comment:
Does FW need this ? i.e. wont we have a custom IPC to load library ? I dont think driver will use this. @plbossart @ujfalusi @ranj063 pls confirm if driver needs this ?

Reply:
I believe we do.. it looks like we are already lagging behind as the kernel is missing `SOF_COMP_MODULE_ADAPTER` type.

Reply:
> I believe we do.. it looks like we are already lagging behind as the kernel is missing `SOF_COMP_MODULE_ADAPTER` type.

What would the kernel use this for ?

Reply:
Can we leave it as is?

Reply:
That line of code fires up ABI accordance error. Can it be as is?

Reply:
please don't add more component type in sof_comp_type which is not supported by ipc4. Please check that there is no copier type.

Reply:
why we need a new module type to load module ? Loading module should be a function of base_fw, so we can do it by base module.

Reply:
Library manager is responsible for loading closed and opensource module. Ipc specific config size is not known for base firmware but library manager must pass this configuration. That is why we need new type of load module to not affect legacy sof mechanizm. Look up [src/ipc/ipc4/helper.c](https://github.com/pjdobrowolski/sof_zephyr/blob/c238559dd7adb88354a9f46e4d6047d71f1c76b0/src/ipc/ipc4/helper.c#L105)

Reply:
@ujfalusi @ranj063 fyi - you may need this on Linux driver. Please review/comment.

Reply:
@pjdobrowolski why do you need to know the IPC size at all? In my PR, I pass the pointer to the mailbox payload as is without the size info https://github.com/thesofproject/sof/pull/5994/commits/80a947da9850113cc13f79ad7742889db6d631ac

Reply:
It works for sof but in legacy platforms payload size would change depends on what module you load.

Reply:
When you load built in module, the base FW exactly knows its configuration size since it has all module headers already. For loadable modules configuration is proprietary and base FW can not figure out how large is the config in IPC buffer. To pass it to the module instance we need to know the size or pass whole 4kB IPC buffer.

Reply:
@jxstelter what is the downside to passing the entire mailbox buffer like in my PR?

Reply:
There are two problems for loadable modules:
1. Memory used - currently module adapter allocates buffer for the configuration and stores it in struct processing_module->priv pointer. We would need to allocate 4kB here. This is waste of memory.
2. Entry point function for IADK loadable modules expects within its parameters pointer to config buffer + size of valid data. 
Already developed modules will not start without correct configuration size.

Reply:
> 1. Memory used - currently module adapter allocates buffer for the configuration and stores it in struct processing_module->priv pointer. We would need to allocate 4kB here. This is waste of memory.

@jxstelter have you seen my PR https://github.com/thesofproject/sof/pull/5994/commits/13cdc38099d5bedbeea5774128cb74080ba8f994? I do no allocate memory for the config but instead just pass the pointer to the IPC payload directly

Reply:
> > 1. Memory used - currently module adapter allocates buffer for the configuration and stores it in struct processing_module->priv pointer. We would need to allocate 4kB here. This is waste of memory.
> 
> @jxstelter have you seen my PR [13cdc38](https://github.com/thesofproject/sof/commit/13cdc38099d5bedbeea5774128cb74080ba8f994)? I do no allocate memory for the config but instead just pass the pointer to the IPC payload directly

OK, this resolves problem with additional allocation. However still we need pass configuration size to IADK modules.

Reply:
@jxstelter do you have an example of where this size of valid data coems from?

Reply:
It is provided in IPC header for ipc4_init_module_instance():
**module.extension.r.param_block_size**